src/components/UriViewer/LogViewer.ts

```diff
@@ -0,0 +1,231 @@
+import { LoadedState } from '@terra-ui-packages/core-utils';
+import _ from 'lodash/fp';
+import { useCallback, useEffect, useState } from 'react';
+import { dd, div, dl, dt, h, span } from 'react-hyperscript-helpers';
+import { ButtonOutline } from 'src/components/common';
+import { centeredSpinner, icon } from 'src/components/icons';
+import Modal from 'src/components/Modal';
+import { InfoBox } from 'src/components/PopupTrigger';
+import { SimpleTabBar } from 'src/components/tabBars';
+import { isAzureUri } from 'src/components/UriViewer/uri-viewer-utils';
+import { Ajax } from 'src/libs/ajax';
+import { useCancellation } from 'src/libs/react-utils';
+import { newTabLinkProps } from 'src/libs/utils';
+/**
+ * Information needed to preview a log file.
+ * @member logUri - The URI of the log file. Must be a valid Azure blob URI. No Sas token should be appended: a fresh one will be obtained.
+ * @member logTitle - The title of the log. Displayed to the user as a tab title.
+ * @member logKey - A unique key for this particular log.
+ * @member logFilename - The filename of this particular log. Does not need to be unique.
+ */
+export type LogInfo = {
+  logUri: string;
+  logTitle: string;
+  logKey: string;
+  logFilename: string;
+};
+
+/**
+ * Props for the LogViewer component.
+ * @member modalTitle - Should represent the group of logs being displayed. E.g. "Task Logs"
+ * @member logs - An array of LogInfo objects.
+ */
+export type LogViewerProps = {
+  modalTitle: string;
+  logs: LogInfo[];
+  onDismiss: () => void;
+};
+
+/**
+ * Props for the SimpleTab component. See src/components/tabBars.js
+ */
+type SimpleTabProps = {
+  key: string;
+  title: string;
+  width: number;
+};
+
+type FetchedLogData = {
+  textContent: string | undefined; // The literal characters of the log file.
+  downloadUri: string | undefined; // The URI to use for downloading the log file. May or may not have token appended, depending on if the file is public or private.
+};
+
+const logLoadingErrorMessage =
+  "Log file could not be loaded. If the workflow or task is still in progress, the log file likely hasn't been generated yet. Some logs may be unavailable if the workflow or task failed before they could be generated.";
+const modalMaxWidth = 1100;
+const tabMaxWidth = modalMaxWidth / 4 - 20;
+
+export const LogViewer = ({ modalTitle, logs, onDismiss }: LogViewerProps) => {
+  const [currentlyActiveLog, setCurrentlyActiveLog] = useState<LogInfo | undefined>(
+    _.isEmpty(logs) ? undefined : logs[0]
+  );
+
+  // string = fetched log content, undefined = loading, null = error.
+  const [activeTextContent, setActiveTextContent] = useState<LoadedState<string | undefined>>({
+    status: 'Loading',
+    state: undefined,
+  });
```

Comment:
```suggestion
  const [activeTextContent, setActiveTextContent] = useState<LoadedState<string>>({
    status: 'Loading',
    state: null,
  });
```

LoadedState allows using null for state that hasn't been loaded. Is there a need to add `undefined` to the type?